### DISCUSSION TITLE
/docs - how to add the token to authorize button?

### DISCUSSION THREAD

**Discussion by adriangallegos** [fastapi/fastapi#8265] [Questions] [answered]
**How to add the token to authorize button?**

First of all I want to congratulate you for the excellent work on FastApi and give my apologize for my english.

I did not find the way to add a Bearer Token to the Authorize Button on "docs page". 

Could you help me? Attach an example

Thanks in advance !!!

Adrián
```python
from fastapi import FastAPI, HTTPException, Security, Depends
from fastapi.security import OAuth2PasswordBearer
from pydantic import BaseModel
from starlette.middleware.cors import CORSMiddleware
import jwt
from datetime import datetime, timedelta

SECRET_KEY = "25cda665298a1ac9a0b5bcf37ddf1c5523f42ccb5cc1f347218b46d42cce7053"
ALGORITHM = "HS256"

class UsuarioGetToken(BaseModel):
    email: str
    password: str = None

class Usuario_Session(BaseModel):
    usuario_id: int
    nombre: str 

class Token(BaseModel):
    token_type: str
    token: str

# FastAPI specific code
app = FastAPI(title='API Test', version='0.1')
app.add_middleware(CORSMiddleware, allow_origins=["*"], allow_methods=["*"], allow_headers=["*"])
token_check = OAuth2PasswordBearer(tokenUrl="/login")    


def get_current_user(token: str = Security(token_check)):
    try:    
        usuario_session = jwt.decode(token, SECRET_KEY, algorithm=ALGORITHM)
    except:
        raise HTTPException(status_code=401, detail="Sessión inválida")    
    return Usuario_Session(**usuario_session)  
    
# Dependency
@app.post("/login", response_model=Token, summary="Autentificación de usuario")
def login(user: UsuarioGetToken):
    to_encode = {"usuario_id": 1, "nombre": user.email}
    expire = datetime.utcnow() + timedelta(minutes=60)
    to_encode.update({"exp": expire, "sub": "Test"})
    access_token = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return Token(token = access_token, token_type = 'Bearer')

@app.get("/test")
def test(current_user: Usuario_Session = Depends(get_current_user)):
    # Verifico los datos
    return {'data': 'ok'}
```

**Comment by euri10** [comment]
the Bearer token is sent as part of the headers
see:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#Authentication_schemes


On Wed, Apr 10, 2019 at 1:44 PM adriangallegos <notifications@github.com>
wrote:

> *How to add the token to authorize button?*
>
> First of all I want to congratulate you for the excellent work on FastApi
> and give my apologize for my english.
>
> I did not find the way to add a Bearer Token to the Authorize Button on
> "docs page".
>
> Could you help me? Attach an example
>
> Thanks in advance !!!
>
> Adrián
>
> from fastapi import FastAPI, HTTPException, Security, Depends
> from fastapi.security import OAuth2PasswordBearer
> from pydantic import BaseModel
> from starlette.middleware.cors import CORSMiddleware
> import jwt
> from datetime import datetime, timedelta
>
>
> SECRET_KEY = "25cda665298a1ac9a0b5bcf37ddf1c5523f42ccb5cc1f347218b46d42cce7053"
> ALGORITHM = "HS256"
>
>
> class UsuarioGetToken(BaseModel):
>
>     email: str
>
>     password: str = None
>
>
> class Usuario_Session(BaseModel):
>
>     usuario_id: int
>
>     nombre: str
>
>
> class Token(BaseModel):
>
>     token_type: str
>
>     token: str
>
>
> # FastAPI specific code
>
> app = FastAPI(title='API Test', version='0.1')
>
> app.add_middleware(CORSMiddleware, allow_origins=["*"], allow_methods=["*"], allow_headers=["*"])
>
> token_check = OAuth2PasswordBearer(tokenUrl="/login")
>
>
>
>
> def get_current_user(token: str = Security(token_check)):
>
>     try:
>
>         usuario_session = jwt.decode(token, SECRET_KEY, algorithm=ALGORITHM)
>
>     except:
>
>         raise HTTPException(status_code=401, detail="Sessión inválida")
>
>     return Usuario_Session(**usuario_session)
>
>
> # Dependency
> @app.post("/login", response_model=Token, summary="Autentificación de usuario")
> def login(user: UsuarioGetToken):
>
>     to_encode = {"usuario_id": 1, "nombre": user.email}
>
>     expire = datetime.utcnow() + timedelta(minutes=60)
>
>     to_encode.update({"exp": expire, "sub": "Test"})
>
>     access_token = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
>
>     return Token(token = access_token, token_type = 'Bearer')
>
>
> @app.get("/test")
> def test(current_user: Usuario_Session = Depends(get_current_user)):
>
>     # Verifico los datos
>
>     return {'data': 'ok'}
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/issues/147>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDZPm7_k2LCUawrh95gqQyh6aXSFBcdks5vfc5_gaJpZM4cmyqU>
> .
>


-- 
benoit barthelet
http://pgp.mit.edu/pks/lookup?op=get&search=0xF150E01A72F6D2EE

**Comment by adriangallegos** [comment]
Hi @euri10, thank you for your response. I use it with postman or insmonia to test the API and works great.

I want to know if is it possible to use on **/docs** with the **try out !!!** option

**Comment by euri10** [comment]
absolutely, if you click on Authorize button and login successfully then the header will be set for you automatically for subsequents api calls with the Try out button

should you use postman you'd have to set it manually like 

```
authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJiZW5vaXQiLCJleHAiOjE1NTQ5MDA5MzksImlzcyI6ImZhc3RhcGlfYXBwZmFjdG9yeSJ9.3mTP3wNIxDXBLnbGW1puwW8k3MWzYOT-VQaO7Z6AdDE
```

**Comment by adriangallegos** [comment]
I tried with the attached example and I get **Auth ErrorError: Bad Request** trying in both client credential locations

**Comment by tiangolo** [comment, marked as answer]
@adriangallegos you have a couple of bugs in the example:

* The JSON containing the token returned MUST have a key `access_token`, `token` alone won't work.
* By declaring `user: UsuarioGetToken` in `login()` you are declaring a JSON body. The OAuth 2.0 spec requires you to receive form data, not JSON, and the fields have to be `username` (not `email`) and `password`.
    * For this, use the provided `OAuth2PasswordRequestForm`, import it and use it as a dependency.
    * Read `OAuth2PasswordRequestForm.username` to get the email of your user. You can use emails as usernames, as long as for authentication you use exactly `username`. That's not a limitation of FastAPI, is part of the spec.

Here's your same code with the minimal changes to work:

```Python
from fastapi import FastAPI, HTTPException, Security, Depends
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from pydantic import BaseModel
from starlette.middleware.cors import CORSMiddleware
import jwt
from datetime import datetime, timedelta

SECRET_KEY = "25cda665298a1ac9a0b5bcf37ddf1c5523f42ccb5cc1f347218b46d42cce7053"
ALGORITHM = "HS256"

class Usuario_Session(BaseModel):
    usuario_id: int
    nombre: str 

class Token(BaseModel):
    token_type: str
    access_token: str

# FastAPI specific code
app = FastAPI(title='API Test', version='0.1')
app.add_middleware(CORSMiddleware, allow_origins=["*"], allow_methods=["*"], allow_headers=["*"])
token_check = OAuth2PasswordBearer(tokenUrl="/login")    


def get_current_user(token: str = Security(token_check)):
    try:    
        usuario_session = jwt.decode(token, SECRET_KEY, algorithm=ALGORITHM)
    except:
        raise HTTPException(status_code=401, detail="Sessión inválida")    
    return Usuario_Session(**usuario_session)  
    
# Dependency
@app.post("/login", response_model=Token, summary="Autentificación de usuario")
def login(form_data: OAuth2PasswordRequestForm = Depends()):
    to_encode = {"usuario_id": 1, "nombre": form_data.username}
    expire = datetime.utcnow() + timedelta(minutes=60)
    to_encode.update({"exp": expire, "sub": "Test"})
    access_token = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return Token(access_token = access_token, token_type = 'Bearer')

@app.get("/test")
def test(current_user: Usuario_Session = Depends(get_current_user)):
    # Verifico los datos
    return {'data': 'ok'}
```

**Comment by adriangallegos** [comment]
Thank you @tiangolo. It's work great !!!

**Comment by tiangolo** [comment]
Great! I'm glad it's working.

Let's then close this issue.

**Comment by ahsanrossi** [comment]
Hi there ,
i am facing this method not allowed error while doing authenticattion

**Reply by ahsanrossi** [reply]
Its working now. Thanks
